harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,471 @@
+                self._collect_metrics()
+                wait_ts = max(next_collection_ts - time.time(), 0)
+                self._should_exit.wait(timeout=wait_ts)
+        finally:
+            self._metrics_queue.put(None)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.datetime.now(datetime.timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
+    ):
+        self._metrics_queue = metrics_queue
+        self._flush_interval = flush_interval
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._metrics: typing.List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        """Start the thread and periodically ship metrics in queue to master.
+
+        This thread batches metrics and reports the batch every `flush_interval` seconds.
+        Currently we make one API call per metric reported, but in the future we should report each
+        batch with one API call.
+        """
+        while True:
+            metrics: typing.List[_Metric] = []
+            next_flush_ts = time.time() + self._flush_interval
+
+            # Pop messages from queue until the next scheduled flush time.
+            while True:
+                now = time.time()
+                timeout = next_flush_ts - now
+
+                if timeout <= 0:
+                    break
+
+                try:
+                    msg = self._metrics_queue.get(timeout=timeout)
+                    if msg is None:
+                        # Received shutdown message, ship accumulated metrics and exit.
+                        self._ship_metrics(metrics=metrics)
+                        return
+
+                    metrics.append(msg)
+                except queue.Empty:
+                    # Timeout reached, no messages in queue.
+                    break
+
+            self._ship_metrics(metrics=metrics)
+
+    def stop(self) -> None:
+        self._metrics_queue.put(None)
+
+    def _ship_metrics(self, metrics: typing.List[_Metric]) -> None:
+        for m in metrics:
+            # Append agent ID to every metric reported.
+            metric = {self._agent_id: m.metrics}
+            self._report_metrics(group=m.group, timestamp=m.timestamp, metrics=metric)
+
+    def _report_metrics(
```

Comment:
Does `_report_metrics` make sense outside the context of `_ship_metrics`? The names are so similar, and anyway the agent annotation and `metrics` types make me wonder if the `_report_metrics` `metrics` is supposed to be a `agent -> _Metric` dict.

It's just a little easier for me to read if I don't also have to think about the semantics of `_report_metrics` as a standalone function.

Reply:
combined this into one function.